File: 0133-clone-graph/0133-clone-graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    // Node* dfs(Node* node, unordered_map<Node*, Node*>& graph2clone) {
    //     if (!node) return nullptr;
    //     if (graph2clone.count(node)) {
    //         return graph2clone[node];
    //     }

    //     Node* curr = new Node(node->val);
    //     graph2clone[node] = curr;

    //     for (Node* nei: node->neighbors) {
    //         Node* nei_clone = dfs(nei, graph2clone);
    //         curr->neighbors.push_back(nei_clone);
    //     }

    //     return curr;
    // }

    // Node* cloneGraph(Node* node) {
    //     unordered_map<Node*, Node*> graph2clone;
    //     return dfs(node, graph2clone);
    // }

    // BFS
    // Time: O(V + E)
    //  - V (# nodes): each node is copied once
    //  - E (# edges): if we pop a node, we loop all of its neighbors. so each edge is processed twice.
    // Space: O(V). including hash map and queue

    Node* cloneGraph(Node* node) {
        if (!node) return nullptr;
        
        unordered_map<Node*, Node*> clones;
        queue<Node*> q;

        clones[node] = new Node(node->val);
        q.push(node);

        while (!q.empty()) {
            int n = q.size();

            for (int i = 0; i < n; i++) {
                Node* curr = q.front();
                q.pop();

                for (Node* nei: curr->neighbors) {
                    if (!clones.count(nei)) {
                        clones[nei] = new Node(nei->val);
                        q.push(nei);
                    }
                    clones[curr]->neighbors.push_back(clones[nei]);
                }
            }
        }

        return clones[node];
    }
};

File: 0133-clone-graph/0133-clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
    # create a copy, add the neighbor of the copy, and return the copy
    def dfs(self, curr, visited_node_map):
        if curr in visited_node_map:
            return visited_node_map[curr]

        visited_node_map[curr] = Node(curr.val)

        for nei in curr.neighbors:
            nei_copy = self.dfs(nei, visited_node_map)
            visited_node_map[curr].neighbors.append(nei_copy)

        return visited_node_map[curr]

    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        if not node:
            return None
        
        visited_node_map = {}
        self.dfs(node, visited_node_map)
        return visited_node_map[node]


# class Solution:
#     def dfs(self, curr, visited_node_map):
#         for nei in curr.neighbors:
#             if nei in visited_node_map:
#                 visited_node_map[curr].neighbors.append(visited_node_map[nei])
#                 continue

#             visited_node_map[nei] = Node(nei.val)
#             visited_node_map[curr].neighbors.append(visited_node_map[nei])            
#             self.dfs(nei, visited_node_map)

#     def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
#         if not node:
#             return None
        
#         visited_node_map = { node: Node(node.val) }
#         self.dfs(node, visited_node_map)

#         return visited_node_map[node]


# from collections import deque
# from typing import Optional
# class Solution:    
#     def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
#         if not node:
#             return None
        
#         q = deque([node])
#         visited_node_map = { node: Node(node.val) }

#         while q:
#             curr = q.popleft()
#             for nei in curr.neighbors:
#                 if nei in visited_node_map:
#                     visited_node_map[curr].neighbors.append(visited_node_map[nei])
#                     continue

#                 visited_node_map[nei] = Node(nei.val)
#                 visited_node_map[curr].neighbors.append(visited_node_map[nei])
#                 q.append(nei)

#         return visited_node_map[node]

"""
test case: graph = 1 - 2, 1 - 3, 2 - 3 (node = 1)
trace:
q = [1]
built = { 1: 1' }
q = [1]
    q = []
    curr = 1
    neighbors
        nei = 2
            built = { 1: 1'(nei=[2']), 2: 2' }
            q = [2]
        nei = 3
            built = { 1: 1'(nei=[2', 3']), 2: 2', 3: 3' }
            q = [2,3]
q = [2,3]
    q = [3]
    curr = 2
    neighbors
        nei = 1
            built = { 1: 1'(nei=[2', 3']), 2: 2'(nei=[1']), 3: 3' }
        nei = 3
            built = { 1: 1'(nei=[2', 3']), 2: 2'(nei=[1', 3']), 3: 3' }
q = [3]
    q = []
    curr = 3
    neighbors
        nei = 1
            built = { 1: 1'(nei=[2', 3']), 2: 2'(nei=[1', 3']), 3: 3'(nei=[1']) }
        nei = 2
            built = { 1: 1'(nei=[2', 3']), 2: 2'(nei=[1', 3']), 3: 3'(nei=[1',2']) }
"""